sphinxcontrib/opencontracting.py

```diff
@@ -211,11 +213,104 @@ def run(self):
+    required_arguments = 1
+    final_argument_whitespace = True
+    option_spec = {'block': directives.unchanged}
+
+    def run(self):
+        title = self.arguments[0]
+        block = self.options.pop('block', '')
+        return [WorkedExampleListNode(block=block, title=title)]
+
+
+class WorkedExample(Directive):
+    required_arguments = 1
+    final_argument_whitespace = True
+    option_spec = {'block': directives.unchanged}
+
+    def run(self):
+        env = self.state.document.settings.env
+        title = self.arguments[0]
+        target_id = f'worked-example-{env.new_serialno("worked-example")}'
+        target_node = nodes.target('', '', ids=[target_id])
+        # A dummy node as we don't want to show anything for the worked example, only mark the content as one
+        ocds_block = self.options.pop('block', '')
+        node = nodes.paragraph('')
+        if not hasattr(env, WORKED_EXAMPLES_ENV_NAME):
+            setattr(env, WORKED_EXAMPLES_ENV_NAME, [])
+
+        getattr(env, WORKED_EXAMPLES_ENV_NAME).append({
+            'docname': env.docname,
+            'lineno': self.lineno,
+            'target': target_node,
+            'block': ocds_block,
+            'title': title,
+        })
+
+        return [target_node, node]
+
+
+def purge_worked_examples(app, env, docname):
+    """
+    Method for removing any existing worked examples from old builds
+    """
+    if not hasattr(env, WORKED_EXAMPLES_ENV_NAME):
+        return
+
+    setattr(env, WORKED_EXAMPLES_ENV_NAME, [worked_example for worked_example in getattr(env, WORKED_EXAMPLES_ENV_NAME)
+                                            if worked_example['docname'] != docname])
+
+
+def process_worked_example_nodes(app, doctree, fromdocname):
+    env = app.builder.env
+
+    if not hasattr(env, WORKED_EXAMPLES_ENV_NAME):
+        setattr(env, WORKED_EXAMPLES_ENV_NAME, [])
+
+    for node in doctree.traverse(WorkedExampleListNode):
+        block = node['block']
+        title = node['title']
+        admonition_node = nodes.admonition('')
+        admonition_node['classes'] += ['admonition', 'note']
+        title = nodes.title('', title)
+        admonition_node += title
+        items = []
+        for worked_example in getattr(env, WORKED_EXAMPLES_ENV_NAME):
+            if block != worked_example['block']:
+                continue
+            uri = app.builder.get_relative_uri(fromdocname,
+                                               f"{worked_example['docname']}#{worked_example['target']['refid']}")
+            reference = nodes.reference('', worked_example['title'], refuri=uri)
+            reference['translatable'] = True
+            paragraph = nodes.paragraph('', '', reference)
+            item = nodes.list_item('', paragraph)
+            items.append(item)
+        admonition_node += nodes.bullet_list('', *items)
+        if not items:
+            raise logging.warning(f'No worked examples are related to {block}')
```

Comment:
Might as well raise before adding an empty list to the bullet list.

Also for formatting log messages: https://ocp-software-handbook.readthedocs.io/en/latest/python/style_guide.html#string-formatting

```suggestion
        if not items:
            raise logging.warning('No worked examples are tagged with %s', block)
        admonition_node += nodes.bullet_list('', *items)
```